### ISSUE TITLE
An unexpected error occurs when conducting RSA by nltools

### ISSUE THREAD

**Issue by huyang200606** [cosanlab/nltools#322]
Dear nltools developers,

    I'm using this wondeful toolbox to perform RSA on my own dataset. Sadly an unexpected error occurred when I performed inter-subject RSA analyses with the following command:
rsa = ROI_dist.similarity(model_dist, plot = False, perm_type = None, metric = 'spearman')

     Then I checked the structure of two variables, i.e., ROI_dist and model_dist. Both variables are 44-by-44 matrices (i.e., 44 participants) so that their dimensions should match with each other. Interestingly, no such error occurs when I perform within-subject RSA using the exact same command.

    I really have no idea what is wrong here. Could you please help me with that? 

    Looking forward to your reply. Thanks a lot for your help!

Best,
Yang

Reply by Luke and Eshin:
Hi Yang,

First, thanks for using our toolbox and providing feedback!

It looks like you need to cast model_dist as an Adjacency object. Make sure they are the same type (e.g., symmetrical similarity or distance matrix)

From your output it looks like model_dist is a 2d numpy array and ROI_dist is a 2d Adjacency. 

When you call ROI_dist.similarity, it’s automatically trying to convert model_dist into and Adjacency for you based on 2 main checks:
1) Is it symmetric? i.e. are the upper triangle elements = lower triangle elements
2) Is it a similarity or distance matrix? Similarity matrices only have 1s on the main diagonal, while distance matrices only have 0s

I can reproduce your error if ROI_dist is symmetrical, but model_dist is not. So I’d recommend converting model_dist to an Adjacency like Luke mentioned and ensuring that both ROI_dist and model_dist are symmetric and are both similarity or distance matrices.

**Comment by huyang200606** [comment]
Thanks for your helpful suggestions! As you exactly pointed out, when I transform the **model_dist** to **Adjacency** variable, it works :-)

However, I still wonder why this error only takes place when I perfom inter-subject rsa (rather than within-subject rsa). A slight difference is that for within-subject RSA, the behavioral RDM is contructed via np.append command, while for inter-subject RSA, the behavioral RDM is constructed via pairwise_distances command. For the rest part both analyses are the same.

**Comment by ejolly** [comment]
Glad it's working. Not sure exactly what the differences are without seeing the data directly. I would make sure in your own analyses code that the matrices you're operating on are indeed symmetric. That's usually the cause of errors when performing RSA analyses square matrices in `nltools` (and other packages).  

If I recall correctly `pairwise_distances` always returns a symmetric 2d array [doc](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise_distances.html), where as `np.append` is just a general purpose function for concatenating arrays of any shape, size, symmetry [doc](https://docs.scipy.org/doc/numpy/reference/generated/numpy.append.html).

**Comment by huyang200606** [comment]
Thanks for your reply and nice suggestions :-) I'll absolutely take care of this issue when performing later RSA.
